### DISCUSSION TITLE
Is VegaFusion usable in Panel/Lumen?

### DISCUSSION THREAD

**Discussion by kszlim** [vega/vegafusion#396] [Q&A] [answered]
Curious if VegaFusion is usable within Holoviz's Panel/Lumen?

**Comment by jonmmease** [comment]
Panel does include some Jupyter Widget support, but last I tried there were issues getting it to work with `VegaFusionWidget`. See https://github.com/bokeh/ipywidgets_bokeh/issues/46#issuecomment-1575747942, and cc [philippjfr](https://github.com/philippjfr).

Would the pre-transform workflow work for your case? This is the approach used by the [VegaFusion mime renderer](https://vegafusion.io/mime_renderer.html) (and by the Altair 5.1's [`"vegafusion"` data transformer](https://altair-viz.github.io/user_guide/large_datasets.html#vegafusion-data-transformer)) where VegaFusion runs once up front to pre-evaluate the transforms that are not part of selection interactions.  If so, we might be able to wire this up using the Altair 5.1 `chart.to_dict(format="vega")` method and the Panel [`Vega` pane](https://panel.holoviz.org/reference/panes/Vega.html).

If you're interested in this I can try out making an example. But if your goal is to support scenarios with Altair selections that interactively filter data, then this approach wouldn't help much. This is the scenario that VegaFusionWidget is designed for.

**Reply by kszlim** [reply]
By "pre-evaluation" do you mean you can only perform fixed transforms ahead of time?

**Reply by jonmmease** [reply, marked as answer]
Yeah. For the example of a histogram ([like this](https://vega.github.io/vega-lite/examples/histogram.html)), VegaFusion will produce a Vega spec that has had binning and aggregation applied, so there will be one row of inline data for each histogram bar. The pre-transformed spec can be rendered by any regular Vega renderer.

But if the spec uses Vega-Lite selections to perform cross-filtering across histograms ([like this](https://vega.github.io/vega-lite/examples/interactive_layered_crossfilter.html)), then it's not possible to evaluate the histogram binning and aggregation up front, because it would break the interactive behavior of the chart. In cases like this, the mime renderer needs to inline the full dataset in the resulting Vega spec.

`VegaFusionWidget` has a special Vega renderer that knows how to send interactive selection events back to the Python kernel, making it possible to avoid sending the full input dataset to the browser while maintaining interactivity. Hope that makes some sense!

**Reply by kszlim** [reply]
Ah, I'm the use case I'd ideally want is something where I can pre-aggregate my dataset when initially viewing it (subsampling it perhaps), and when zooming in I want the full resolution data. I'm guessing that use case probably doesn't work right now?

**Reply by jonmmease** [reply]
To do resampling on zoom we'd need the full functionality of `VegaFusionWidget` to be available.

FWIW, `VegaFusionWidget` does work with Voila and Solara.
